### PR TITLE
PP-1603 Invite Resource OTP

### PR DESCRIPTION
**Pull request by DanailMinchev**
## WHAT

- Add invites expiry date. So move logic into the entity.
- Refactored invites otp resources paths.
- Add validation for otp requests.
- Add resource to create user on otp validation.
- Add resource to resend otp
- Return newly created User object on successful invite validation

### CODE REVIEW

File: src/main/java/uk/gov/pay/adminusers/service/InviteService.java [hunk 1]
```diff
@@ -131,7 +134,21 @@ public void generateOtp(InviteOtpRequest inviteOtpRequest) {
...
         } else {
-            throw notFoundException();
+            throw notFoundInviteException(inviteOtpRequest.getCode());
+        }
+    }
+
+    @Transactional
+    public User createInvitedUser(InviteValidateOtpRequest inviteValidateOtpRequest) {
+        Optional<InviteEntity> inviteOptional = inviteDao.findByCode(inviteValidateOtpRequest.getCode());
+        if (inviteOptional.isPresent() && secondFactorAuthenticator.authorize(inviteOptional.get().getOtpKey(), inviteValidateOtpRequest.getOtpCode())) {
```
Comment (kakumara) [inline src/main/java/uk/gov/pay/adminusers/service/InviteService.java]: How you going identify between if the invite was not found or the second factor authentication error? As discussed selfservice needs to know the differentiation in order to direct user either to re-enter otp or stop registration altogether.

Comment (kakumara) [inline src/main/java/uk/gov/pay/adminusers/service/InviteService.java]: How you going identify between if the invite was not found or the second factor authentication error? As discussed selfservice needs to know the differentiation in order to direct user either to re-enter otp or stop registration altogether.

File: src/test/java/uk/gov/pay/adminusers/resources/InviteResourceOtpTest.java [hunk 2]
```diff
@@ -2,58 +2,218 @@
...
                 .build();
 
         givenSetup()
                 .when()
                 .body(mapper.writeValueAsString(invitationRequest))
                 .contentType(ContentType.JSON)
-                .post(format(INVITES_OTP_RESOURCE_URL, "not-existing-code"))
+                .post(INVITES_GENERATE_OTP_RESOURCE_URL)
+                .then()
+                .statusCode(BAD_REQUEST.getStatusCode());
```
Comment (kakumara) [inline src/test/java/uk/gov/pay/adminusers/resources/InviteResourceOtpTest.java]: Not exist returns BAD_REQUEST?  shouldn't it be NOT_FOUND ?

File: src/test/java/uk/gov/pay/adminusers/resources/InviteResourceOtpTest.java [hunk 3]
```diff
@@ -2,58 +2,218 @@
...
+
+        // check if the user has been created and it is not disabled in the database
+        List<Map<String, Object>> users = databaseHelper.findUserByUsername(EMAIL);
+
+        assertThat(users.size(), is(1));
+
+        Map<String, Object> createdUser = users.get(0);
+        assertThat(createdUser.get("username"), is(EMAIL));
+        assertThat(createdUser.get("email"), is(EMAIL));
+        assertThat(createdUser.get("password"), is(PASSWORD));
```
Comment (kakumara) [inline src/test/java/uk/gov/pay/adminusers/resources/InviteResourceOtpTest.java]: Saving plain test password ???? this should be encrypted using encryptor.. otherwise login will not work. Have a look at UserService.java

Comment (kakumara) [inline src/test/java/uk/gov/pay/adminusers/resources/InviteResourceOtpTest.java]: Instead of creating a UserEntity directly from InviteEntity .. i would rather create a User.java from the invite and pass it to UserService to save (so that it could do the relevant validations and encryptions there)

File: src/test/java/uk/gov/pay/adminusers/resources/InviteResourceOtpTest.java [hunk 4]
```diff
@@ -2,58 +2,218 @@
...
+                .put("otp", PASSCODE)
+                .build();
+
+        givenSetup()
+                .when()
+                .body(mapper.writeValueAsString(invitationOtpRequest))
+                .contentType(ContentType.JSON)
+                .post(INVITES_VALIDATE_OTP_RESOURCE_URL)
+                .then()
+                .statusCode(BAD_REQUEST.getStatusCode());
```
Comment (kakumara) [inline src/test/java/uk/gov/pay/adminusers/resources/InviteResourceOtpTest.java]: Once again non-existent code becomes BAD_REQUEST? shouldn't it be NOT_FOUND?

File: src/test/java/uk/gov/pay/adminusers/resources/InviteResourceOtpTest.java [hunk 5]
```diff
@@ -2,58 +2,218 @@
...
+                .put("otp", 123456)
+                .build();
+
+        givenSetup()
+                .when()
+                .body(mapper.writeValueAsString(invitationOtpRequest))
+                .contentType(ContentType.JSON)
+                .post(INVITES_VALIDATE_OTP_RESOURCE_URL)
+                .then()
+                .statusCode(BAD_REQUEST.getStatusCode());
```
Comment (kakumara) [inline src/test/java/uk/gov/pay/adminusers/resources/InviteResourceOtpTest.java]: BAD_REQUST? Shouldn't this be a 401 (UNAUTHORIZED) ?

File: src/main/java/uk/gov/pay/adminusers/model/User.java [hunk 6]
```diff
@@ -18,10 +18,8 @@
     public static final String FIELD_USERNAME = "username";
     public static final String FIELD_PASSWORD = "password";
     public static final String FIELD_EMAIL = "email";
-    public static final String FIELD_SERVICE_IDS = "service_ids";
     public static final String FIELD_GATEWAY_ACCOUNT_IDS = "gateway_account_ids";
     public static final String FIELD_TELEPHONE_NUMBER = "telephone_number";
-    public static final String FIELD_OTP_KEY = "otp_key";
```
Comment (kakumara) [inline src/main/java/uk/gov/pay/adminusers/model/User.java]: Why are these removed from User?

File: src/main/java/uk/gov/pay/adminusers/persistence/entity/InviteEntity.java [hunk 7]
```diff
@@ -144,27 +175,25 @@ public Invite toInvite() {
...
-     * - Invite is not expired because 02-05-2017 17:50:10 < 03-05-2017 00:00:00
-     *
-     * @return true if expired
-     */
     public boolean isExpired() {
-        return now(UTC).isAfter(date.truncatedTo(DAYS).plus(EXPIRY_DAYS, DAYS));
+        return now(UTC).isAfter(expiryDate);
+    }
+
+    public UserEntity mapToUserEntity() {
```
Comment (kakumara) [inline src/main/java/uk/gov/pay/adminusers/persistence/entity/InviteEntity.java]: See comment below. I think it would be better to create a User.java instead of UserEntity here..

File: src/main/java/uk/gov/pay/adminusers/service/InviteService.java [hunk 8]
```diff
@@ -48,15 +48,18 @@ public InviteService(RoleDao roleDao,
                          InviteDao inviteDao,
                          AdminUsersConfig config,
                          PasswordHasher passwordHasher,
-                         NotificationService notificationService, SecondFactorAuthenticator secondFactorAuthenticator) {
+                         NotificationService notificationService,
```
Comment (kakumara) [inline src/main/java/uk/gov/pay/adminusers/service/InviteService.java]: This constructor is growing and have the same symptoms to that of UserService. I would have preferred an approach similar to ServiceFactory approach (see UserServicesFactory.java) so that the business logic can scale horizontally.
Although I'm not too picky on this comment and happy to do this kind of refactoring at a later stage if its too much

File: src/main/java/uk/gov/pay/adminusers/service/InviteService.java (hunk 1 above)
Comment (DanailMinchev) [inline src/main/java/uk/gov/pay/adminusers/service/InviteService.java]: I have added `401 - UNAUTHORIZED` with the log message `Invite for code %s provided invalid otp auth code`

**Comment by kakumara** [comment]
Also it seems you are missing `@Transactional` in InviteService.generateOtp()

File: src/main/java/uk/gov/pay/adminusers/model/User.java (hunk 6 above)
Comment (DanailMinchev) [inline src/main/java/uk/gov/pay/adminusers/model/User.java]: Because `FIELD_OTP_KEY ` is not used anywhere, let me know if you want me to add it again.

File: src/test/java/uk/gov/pay/adminusers/resources/InviteResourceOtpTest.java (hunk 3 above)
Comment (DanailMinchev) [inline src/test/java/uk/gov/pay/adminusers/resources/InviteResourceOtpTest.java]: Thanks for spotting that!

File: src/main/java/uk/gov/pay/adminusers/service/InviteService.java (hunk 8 above)
Comment (DanailMinchev) [inline src/main/java/uk/gov/pay/adminusers/service/InviteService.java]: As discussed this will be part of refactoring story

File: src/test/java/uk/gov/pay/adminusers/resources/InviteResourceOtpTest.java (hunk 3 above)
Comment (DanailMinchev) [inline src/test/java/uk/gov/pay/adminusers/resources/InviteResourceOtpTest.java]: Actually, the InviteEntity is already being used with the hashed password when it is created: https://github.com/alphagov/pay-adminusers/blob/master/src/main/java/uk/gov/pay/adminusers/service/InviteService.java#L121 and we are inserting the `PASSWORD` (`a-secure-password`) as a `InviteEntity` fixture

Comment (DanailMinchev) [inline src/test/java/uk/gov/pay/adminusers/resources/InviteResourceOtpTest.java]: About the password comment:

Actually the `InviteEntity` is already being used with hashed password when it is created: https://github.com/alphagov/pay-adminusers/blob/master/src/main/java/uk/gov/pay/adminusers/service/InviteService.java#L121 and we are inserting the PASSWORD (a-secure-password) as a InviteEntity fixture, that is why we are comparing directly.

About creating the `UserEntity` directly, as discussed we will need to refactor this later, because we have only the encrypted password from `InviteEntity` and `User` or `CreateUserRequest` needs plain text version

File: src/main/java/uk/gov/pay/adminusers/persistence/entity/InviteEntity.java (hunk 7 above)
Comment (DanailMinchev) [inline src/main/java/uk/gov/pay/adminusers/persistence/entity/InviteEntity.java]: Please see my comment above..

File: src/test/java/uk/gov/pay/adminusers/resources/InviteResourceOtpTest.java (hunk 3 above)
Comment (kakumara) [inline src/test/java/uk/gov/pay/adminusers/resources/InviteResourceOtpTest.java]: 👍

File: src/test/java/uk/gov/pay/adminusers/resources/InviteResourceOtpTest.java (hunk 2 above)
Comment (kakumara) [inline src/test/java/uk/gov/pay/adminusers/resources/InviteResourceOtpTest.java]: I don't think these 2 tests are really necessary (just one is enough). As discussed
